srv6: Introduce IPv6 address in segment-list used by TE Policy

dmytroshytyi opened the pull request:
These series of patches introduce new feature: user may configure ipv6-address in segment list.

Format:
```
index INDEX ipv6-address X:X::X:X
```

Example:
```
segment-routing
 traffic-eng
  segment-list SL1
   index 0 ipv6-address 2001:db8::1
```

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
zebra: add srv6 nexthop resolution given by sr policy

Consider an srv6 policy active, when sr policy set is
received. The sr_policy_validate() function is however
called, the default state value is down, and immediately
sent to UP.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
zebra: zapi, at least one label or srv6 sid check

SR-TE tunnel must contain at least one label or SRv6 SID.
Check this condition.

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
zebra: fill nexthop with srv6 segs same as for labels

When the segment list is not empty add SRv6 segments
to a nexthop.

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

pguibert6WIND committed:
pathd, lib: add srv6 policy with resolved nexthop info

The Pathd daemon is aware of the SID reachability, but
still once the sr policy installed in zebra, there is need
to re-resolve the SID reachability.

Extend the Pathd NHT service, by extracting the resolved
next-hop. This information will be conveyed in the SR_POLICY_SET
message, and will be very helpful to simplify code in zebra.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
lib,pathd,zebra: fill segment_list with sr types

Add structute sr_types_t, that contains the Segment Routing types.
Fill the zapi_sr_policy's segment_list with values such as:
- value
- local_label
- num_segs
- srv6_segs

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
zclient: add srv6 segs to zapi_srte_tunnel

Add SRv6 segments to zapi_srte_stunnel structure.

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
zebra: encode, decode srv6 segs

Encode and Decode SRv6-TE. It includes:
- number of segments
- list of segments

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
zebra: change srv6 policy based on segs change.

When SRv6 list of segments was changed, change SRv6 policy.

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

pguibert6WIND committed:
doc: add pathd srv6 segment-list support

Add pathd srv6 segment list support.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
topotest: isis_srv6_te_topo1

SRv6 Policy is an ordered list of segments that represent a
source-routed policy. Packet flows are steered into an SRv6
Policy.

Signed-off-by: Dmytro Shytyi <dmytro@shytyi@6wind.com>
Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

dmytroshytyi-6WIND commented:
Fix warnings by checkpatch.

dmytroshytyi-6WIND commented:
2 small fixes for  isis_sr_te uploaded in the recent push.

ton31337 commented on pathd/path_zebra.c:
```diff
@@ -155,32 +336,125 @@ static int path_zebra_router_id_update(ZAPI_CALLBACK_ARGS)
+	if (!was_zebra_registered)
+		return;
+
+	zlog_debug("Unregistering nexthop(%pFX) for candidate %s pref %u",
```
Should be guarded by debug flag?

dmytroshytyi-6WIND replied:
could you please elaborate? It seems like other parts of code do it in similar way? https://github.com/FRRouting/frr/blob/57fdef4550c601fc84be1f2a04e591ea6e19f381/pathd/path_zebra.c#L672
https://github.com/FRRouting/frr/blob/57fdef4550c601fc84be1f2a04e591ea6e19f381/pathd/path_zebra.c#L692

pguibert6WIND replied:
In a future pull request, a 'debug pathd nth' extension command should be planned.

donaldsharp replied:
yeah this is a blocker from my perspective adding more unguarded debugs to a PR, even if the original code snuck some in, is a no go.

dmytroshytyi-6WIND replied:
remove the blocker -> fixed.

ton31337 commented on pathd/path_zebra.c:
```diff
@@ -259,6 +546,109 @@ static void path_zebra_label_manager_connect(void)
+		if (!candidate)
+			continue;
+
+		segment_list = candidate->lsp->segment_list;
```
`candidate->lsp` can be NULL, right?

dmytroshytyi-6WIND replied:
fixed, thanks

ton31337 commented on pathd/path_zebra.c:
```diff
@@ -259,6 +546,109 @@ static void path_zebra_label_manager_connect(void)
+	nhtd = path_nht_hash_find(path_nht_hash, &lookup);
+
+	if (!nhtd)
+		zlog_err("No nhtd?");
```
Something more meaningful maybe?

dmytroshytyi-6WIND replied:
fixed

ton31337 commented on tests/topotests/isis_srv6_te_topo1/rt1/zebra.conf:
```diff
@@ -0,0 +1,41 @@
+! debug zebra packet
+!
+interface lo
+ ip address 1.1.1.1/32
```
Ditto.

dmytroshytyi-6WIND replied:
fixed

ton31337 commented on tests/topotests/isis_srv6_te_topo1/rt2/zebra.conf:
```diff
@@ -0,0 +1,33 @@
+! debug zebra packet
+!
+interface lo
+ ip address 2.2.2.2/32
```
Ditto.

dmytroshytyi-6WIND replied:
fixed

ton31337 commented on tests/topotests/isis_srv6_te_topo1/rt3/zebra.conf:
```diff
@@ -0,0 +1,32 @@
+! debug zebra packet
+!
+interface lo
+ ip address 3.3.3.3/32
```
Ditto.

dmytroshytyi-6WIND replied:
fixed

ton31337 commented on tests/topotests/isis_srv6_te_topo1/rt4/zebra.conf:
```diff
@@ -0,0 +1,35 @@
+! debug zebra packet
+!
+interface lo
+ ip address 4.4.4.4/32
```
Ditto.

dmytroshytyi-6WIND replied:
fixed

ton31337 commented on tests/topotests/isis_srv6_te_topo1/test_isis_srv6_te_topo1.py:
```diff
@@ -0,0 +1,476 @@
+    create_interface_in_kernel,
+)
+
+pytestmark = [pytest.mark.isisd, pytest.mark.sharpd]
```
Shouldn't bgpd, pathd be included also?

dmytroshytyi-6WIND replied:
fixed

donaldsharp replied:
this is not fixed.

dmytroshytyi-6WIND replied:
really fixed by also adding the 
``
pytest.mark.pathd
```

riw777 reviewed (commented):
some questions; waiting on @ton31337 's comments as well

donaldsharp commented on lib/srte.h:
```diff
@@ -19,6 +19,23 @@ enum zebra_sr_policy_status {
 };
 
+/* SR types. */
+enum sr_types_t {
```
Typically XXX_t refers to a typedef.  This is not a typedef.   This should be `enum sr_types` or `enum sr_types_enum`

dmytroshytyi-6WIND replied:
fixed to enum sr_types

donaldsharp commented on lib/srte.h:
```diff
@@ -19,6 +19,23 @@ enum zebra_sr_policy_status {
+	ZEBRA_SR_SRV6_SRTE = 2, /* SRv6 SID List*/
+};
+
+static inline enum lsp_types_t lsp_type_from_sr_type(int sr_type)
```
Shouldn't this pass in a `enum sr_types..` instead of an int? Let's let the compiler help us please

dmytroshytyi-6WIND replied:
fixed

donaldsharp replied:
Please remove the default case here.

dmytroshytyi-6WIND replied:
fixed

donaldsharp commented on pathd/path_zebra.c:
```diff
@@ -336,6 +642,7 @@ static zclient_handler *const path_handlers[] = {
 	[ZEBRA_SR_POLICY_NOTIFY_STATUS] = path_zebra_sr_policy_notify_status,
 	[ZEBRA_ROUTER_ID_UPDATE] = path_zebra_router_id_update,
 	[ZEBRA_OPAQUE_MESSAGE] = path_zebra_opaque_msg_handler,
+	[ZEBRA_NEXTHOP_UPDATE] = path_zebra_nexthop_update,
```
Let's use `zclient->nexthop_update` instead.

dmytroshytyi-6WIND replied:
thanks

donaldsharp commented on pathd/path_zebra.c:
```diff
@@ -284,6 +526,70 @@ static void path_zebra_label_manager_connect(void)
+	}
+}
+
+static int path_zebra_nexthop_update(ZAPI_CALLBACK_ARGS)
```
Please use the `zclient->nexthop_update` call back function instead.  I don't want to remember to have to change pathd when I add/remove data the encode/decode of nexthops

dmytroshytyi-6WIND replied:
thanks

donaldsharp commented on zebra/rt_netlink.c:
```diff
@@ -1550,7 +1550,7 @@ static ssize_t fill_seg6ipt_encap(char *buffer, size_t buflen,
 
 	for (i = 0; i < segs->num_segs; i++) {
-		memcpy(&srh->segments[i], &segs->seg[i],
+		memcpy(&srh->segments[segs->num_segs - i - 1], &segs->seg[i],
```
This is only 1/2 of the equation, we need to fix the decode side as well.  Add the ipv6 seg6 route from the linux cli and the segments are backwards in comparison.

dmytroshytyi-6WIND replied:
Thanks, fixed,

donaldsharp commented on pathd/path_zebra.c:
```diff
@@ -665,6 +723,8 @@ void path_zebra_init(struct event_loop *master)
 	zclient_sync->instance = 1;
 	zclient_sync->privs = &pathd_privs;
 
+	zclient->nexthop_update = path_zebra_nexthop_update;
```
I'm not really interested in having a code changes that uses [ZEBRA_NEXTHOP_UPDATE] and then another commit switches this out.  Let's consolidate and shrink this down to 1 commit.

dmytroshytyi-6WIND replied:
Does it look good in the recent push?

donaldsharp commented on zebra/zebra_srte.c:
```diff
@@ -133,6 +136,48 @@ static int zebra_sr_policy_notify_update_client(struct zebra_sr_policy *policy,
+
+		for (rn = route_top(table); rn; rn = route_next(rn)) {
+			RNODE_FOREACH_RE_SAFE (rn, re, next) {
+				if (re->nhe->nhg.nexthop->srte_color ==
```
invert this logic so the indent is not so deep  this is hard to read

dmytroshytyi-6WIND replied:
fixed.

donaldsharp commented on zebra/zebra_srte.c:
```diff
@@ -133,6 +136,47 @@ static int zebra_sr_policy_notify_update_client(struct zebra_sr_policy *policy,
 
+	if (policy->segment_list.srv6_segs.num_segs > 0) {
+		/* Lookup table.  */
+		table = zebra_vrf_table(AFI_IP6, SAFI_UNICAST, VRF_DEFAULT);
```
What is supposed to happen when we have > 1 million routes?  What is a reasonable number of items that are going to be sent.  Are we going to run out of packet space?  Are we going to just cpuhog like you would not believe?

dmytroshytyi-6WIND replied:
It is not easy to predict what exactly might happen. I will bring a possible improvement, working on that.

donaldsharp commented on tests/topotests/isis_srv6_te_topo1/test_isis_srv6_te_topo1.py:
```diff
@@ -0,0 +1,476 @@
+    assert diff is None, assertmsg
+
+
+def check_ping6(name, dest_addr, expect_connected):
```
Yeah Adding yet *ANOTHER* check_ping6 is a no go from my perspective especially since lib/checkping.py exists.

dmytroshytyi-6WIND replied:
fixed

donaldsharp reviewed (changes requested):
Some work needs to be done to clean this up

dmytroshytyi-6WIND commented:
Thank you very much for your reviews, @ton31337, @riw777, @donaldsharp.

I was on PTO, thus the response delay. I expect to look at this PR in coming time.

@cscarpitta, WDYT about the cli format for ipv6 address? Does it look good?

dmytroshytyi-6WIND commented on tests/topotests/isis_srv6_te_topo1/dst/zebra.conf:
```diff
@@ -0,0 +1,23 @@
+! debug zebra mpls
+!
+interface lo
+ ip address 9.9.9.2/32
```
fixed

dmytroshytyi-6WIND commented on tests/topotests/isis_srv6_te_topo1/rt1/isisd.conf:
```diff
@@ -0,0 +1,35 @@
+password 1
```
fixed

dmytroshytyi-6WIND commented on tests/topotests/isis_srv6_te_topo1/rt1/step1/show_ip_route.ref:
```diff
@@ -0,0 +1,276 @@
+{
+  "2.2.2.2\/32":[
```
fixed

dmytroshytyi-6WIND commented on zebra/zebra_srv6.c:
```diff
@@ -409,6 +409,50 @@ int release_daemon_srv6_locator_chunks(struct zserv *client)
+			       &policy->segment_list.srv6_segs.segs[0]);
+
+		/* Lookup table.  */
+		table = zebra_vrf_table(AFI_IP6, SAFI_UNICAST, VRF_DEFAULT);
```
fixed

dmytroshytyi-6WIND commented on zebra/zebra_srv6.c:
```diff
@@ -409,6 +409,50 @@ int release_daemon_srv6_locator_chunks(struct zserv *client)
+			}
+		}
+	}
+	if (srv6_sid_resolved)
```
fixed

dmytroshytyi-6WIND commented on zebra/zebra_srte.c:
```diff
@@ -133,6 +136,47 @@ static int zebra_sr_policy_notify_update_client(struct zebra_sr_policy *policy,
+						stream_putw(s,
+							    0); /* instance - not available */
+						stream_putc(s,
+							    150); /* TODO: distance */
```
Which distance would you suggest?

dmytroshytyi-6WIND commented on pathd/path_zebra.c:
```diff
@@ -548,40 +568,82 @@ static void path_nht_update(struct prefix *nh, int nhnum)
+	struct nexthop *nhlist_tail = NULL;
+	struct nexthop *oldnh;
+	bool nh_changed = false;
+	char buf[NEXTHOP_STRLEN];
```
fixed

dmytroshytyi-6WIND commented on lib/zclient.c:
```diff
@@ -3711,6 +3711,19 @@ int zapi_sr_policy_encode(struct stream *s, int cmd, struct zapi_sr_policy *zp)
+
+	stream_putw(s, zt->srv6_segs.num_segs);
+
+	stream_put(s, &zt->srv6_segs.segs[0],
```
nice catch, thanks. fixed in encode and decode functions.

dmytroshytyi-6WIND commented on lib/zclient.c:
```diff
@@ -3741,6 +3767,33 @@ int zapi_sr_policy_decode(struct stream *s, struct zapi_sr_policy *zp)
 		STREAM_GETL(s, zt->labels[i]);
 
+	/* Decode SRv6-TE */
+	STREAM_GETW(s, zt->srv6_segs.num_segs);
```
nice catch, this num_segs could be = 0. and in this case we shouldn't decode the segment list. fixed

dmytroshytyi-6WIND commented on tests/topotests/isis_srv6_te_topo1/rt1/step1/show_ipv6_route.ref:
```diff
@@ -0,0 +1,246 @@
+{
+    "fc00:0:1::/128": [
```
This ULA format looks good, these addresses are not used on the internet (unlike 1.1.1.1 with ipv4 that might be configured on the internet), changing it will involve the redefining the topotest from entirely with recreating the jsons references...

dmytroshytyi committed:
zebra: fix srv6 segment-list sid order when read from kernel

When configuring an SRv6 SID list via iproute2, the segment list
that appears in vtysh, learned from the kernel, is inversed.

Fix the order of SRv6 SIDs in segment list, learned from the
kernel.

Fixes: f20cf14 ("bgpd,lib,sharpd,zebra: srv6 introduce multiple
segs/SIDs in nexthop")
Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi-6WIND commented:
ci:rerun

dmytroshytyi-6WIND commented:
Dear @ton31337, @riw777, @donaldsharp. Thank you for reviews. I have responded on all comments and provided a new code version that addresses the comments. It passes all basics sanity checks and topotests. Could you please have a look at this PR again?

pguibert6WIND commented on pathd/path_zebra.c:
```diff
@@ -30,13 +31,93 @@ struct zclient *zclient;
 /* Global Variables */
-bool g_has_router_id_v4 = false;
-bool g_has_router_id_v6 = false;
+bool g_has_router_id_v4;
```
why is it needed to modify those global variables ?

dmytroshytyi-6WIND replied:
fixed

pguibert6WIND commented on pathd/path_zebra.c:
```diff
@@ -558,40 +578,79 @@ static void path_nht_update(struct prefix *nh, int nhnum)
-	lookup.nh = matched;
-	lookup.nh_vrf_id = vrf_id;
+	prefix_copy(&lookup.nh, match);
+	lookup.nh_vrf_id = vrf->vrf_id;
```
to make consistency with each commit, the code changes in path_zebra_nexthop_update() should be moved to the previous "pathd: add registration mechanism"

dmytroshytyi-6WIND replied:
fixed

pguibert6WIND commented on zebra/zebra_srte.c:
```diff
@@ -81,6 +81,60 @@ struct zebra_sr_policy *zebra_sr_policy_find_by_name(char *name)
+					stream_putc(s, ZEBRA_ROUTE_SRTE);
+					stream_putw(s,
+						    0); /* instance - not available */
+					stream_putc(s, 150); /* TODO: distance */
```
replace 150 with ZEBRA_SHARP_DISTANCE_DEFAULT (defined in zebra.h)
if there is a problem with the given distance value, changes should be done in zebra.h, now.

dmytroshytyi-6WIND replied:
fixed

pguibert6WIND reviewed (commented):
some additional commentS.

riw777 reviewed (approved):
looks good ... waiting on comments from other reviewers

dmytroshytyi committed:
pathd: add registration mechanism

Pathd needs to track the reachability of the first SID
of any SRv6 segment-list. Use the nexthop tracking
facility and add the pathd daemon as a client to nht service
from zebra.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi-6WIND commented:
ci:rerun

pguibert6WIND commented on tests/topotests/isis_srv6_te_topo1/test_isis_srv6_te_topo1.py:
```diff
@@ -0,0 +1,472 @@
+        netmask="128",
+        create=True,
+    )
+    create_interface_in_kernel(
```
this would be nice to really have the sr0 interface active.
for instance, use the following call for each device.
```
    shutdown_bringup_interface_in_kernel(
        tgen,
        "rt2",
        "sr0",
        ifaceaction=True
    )
```

pguibert6WIND reviewed (commented):
additional remark

pguibert6WIND committed:
pathd, lib, zebra: srte, add distance and metric values in srv6 segment_list

When notifying an update for a colored nexthop, ZEBRA sends the default PATHD
distance value and the 0 metric value. Those two values break the original
values that originated from an IGP route, for instance. When transmitting this
information to BGP for instance, the distance and values are very important
when computing BGP paths. Change this by keeping the original distance and
metric values.

When SRv6 segment list size is bigger than 0, send segments
and other elements via zserv_send_message().

The ZEBRA SRTE code can not rely on the routing table. Use instead the
resolved nexthop information gathered by th SRv6 SRTE information.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

mwinter-osr commented:
CI:rerun libyang version may have changed during execution which caused the failure

dmytroshytyi-6WIND commented:
@donaldsharp, you raised the next question:
```
What is supposed to happen when we have > 1 million routes? What is a reasonable number of items that are going to be sent. Are we going to run out of packet space? Are we going to just cpuhog like you would not believe?
```
There is no more a loop  over all routes and the max segs is verified and limited if exceeds the defined in header value.
Thanks.

dmytroshytyi-6WIND commented:
ci:eerun

dmytroshytyi-6WIND commented:
ci:rerun

dmytroshytyi-6WIND commented:
Hello, @donaldsharp @ton31337 @riw777. The PR passes the CI tests and comments are addressed.
WDYT?

riw777 commented:
looks good to me ... waiting on @donaldsharp, I think?

dmytroshytyi-6WIND commented:
> looks good to me ... waiting on @donaldsharp, I think?

I think so, seems like the blocker (changes requested) was introduced  by @donaldsharp. It would be perfect to get his reply based on the latest status of PR.

dmytroshytyi-6WIND commented:
Rebase was performed.

dmytroshytyi-6WIND commented:
ci:rerun

dmytroshytyi-6WIND commented:
ci:rerun
unrelated failures

dmytroshytyi-6WIND commented:
ci:rerun

dmytroshytyi-6WIND commented:
Hello @ton31337 , @donaldsharp, @riw777 
Thank you very much for these reviews.
Is there anything you would like to add as a comments for this PR?

dmytroshytyi-6WIND commented:
ci-rerun

dmytroshytyi-6WIND commented:
ci:rerun

dmytroshytyi-6WIND commented:
Hello @donaldsharp,

Basic tests are OK.

Could you please take a look at this PR?

Thanks in advance.

dmytroshytyi-6WIND commented:
Hello @riw777 @ton31337 @donaldsharp
1. basic tests are OK
2. There is old "changes requested" that blocks the merging. All comments already were addressed some time ago.

@donaldsharp, no more new comments from reviewers for some time...  If you don't have any comments left, please change the status of your review. Hopefully we will be able to unblock and merge it.

riw777 commented:
I think the only thing holding this is @donaldsharp clearing his change request block ...

donaldsharp commented on lib/srte.h:
```diff
@@ -19,6 +19,25 @@ enum zebra_sr_policy_status {
+		return ZEBRA_LSP_SRTE;
+	case ZEBRA_SR_LSP_NONE:
+	case ZEBRA_SR_SRV6_SRTE:
+	default:
```
remove the default this is an enum and all the values should be accounted for in the switch statement.  Leaving a default allows future changes to the enum to be masked here and the developer will miss it.

dmytroshytyi-6WIND replied:
fixed

donaldsharp commented on pathd/pathd.c:
```diff
@@ -621,7 +622,6 @@ void srte_policy_apply_changes(struct srte_policy *policy)
 			SET_FLAG(new_best_candidate->flags, F_CANDIDATE_BEST);
 			SET_FLAG(new_best_candidate->flags,
 				 F_CANDIDATE_MODIFIED);
-
```
why the whitespace change?

dmytroshytyi-6WIND replied:
fixed

donaldsharp commented:
When I run the new test under any type of load it fails badly:

```
================================================================== short test summary info ==================================================================
FAILED ospf6_ecmp_inter_area/test_h.py::test_ecmp_inter_area - AssertionError: 'r1' wrong number of route nexthops
FAILED ospf6_ecmp_inter_area/test_e.py::test_ecmp_inter_area - AssertionError: 'r1' wrong number of route nexthops
FAILED isis_srv6_te_topo1/test_zy.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zs.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zd.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_ze.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zz.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zr.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zi.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zc.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zh.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zv.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_isis_srv6_te_topo1.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zt.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zn.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zm.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zk.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zo.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zw.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zj.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zb.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zu.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_za.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zp.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zf.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zl.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
FAILED isis_srv6_te_topo1/test_zq.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
FAILED ospf6_ecmp_inter_area/test_b.py::test_ecmp_inter_area - AssertionError: 'r1' wrong number of route nexthops
FAILED isis_srv6_te_topo1/test_zx.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
================================================= 29 failed, 1732 passed, 279 skipped in 819.59s (0:13:39) =============================```

Please spend some time running your new test under different loads and identifying and fixing the problems found.

donaldsharp commented:
as a note when I run the test by itself or concurrently with itself it works fine.  It's only when scale/load is placed on the system that causes it to fail badly.  This means, to me, that the test is starting to look for data before some other event has occurred and we get this output.  See the hundreds of fixes in the topotest directory as we fix this sort of problem in the test for guidance on how to approach the problem.  #16586, #16583, #16523, #16510 and #16485 as recent examples of this sort of problem.

dmytroshytyi-6WIND commented:
> When I run the new test under any type of load it fails badly:
> 
> ```
> ================================================================== short test summary info ==================================================================
> FAILED ospf6_ecmp_inter_area/test_h.py::test_ecmp_inter_area - AssertionError: 'r1' wrong number of route nexthops
> FAILED ospf6_ecmp_inter_area/test_e.py::test_ecmp_inter_area - AssertionError: 'r1' wrong number of route nexthops
> FAILED isis_srv6_te_topo1/test_zy.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zs.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zd.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_ze.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zz.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zr.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zi.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zc.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zh.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zv.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_isis_srv6_te_topo1.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zt.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zn.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zm.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zk.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zo.py::test_srv6_te_policy_removed - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zw.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zj.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zb.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zu.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_za.py::test_rib_ipv6_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zp.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zf.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zl.py::test_srv6_te_policy_activated - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED isis_srv6_te_topo1/test_zq.py::test_srv6_locator_step1 - AssertionError: "rt1" JSON output mismatches the expected result
> FAILED ospf6_ecmp_inter_area/test_b.py::test_ecmp_inter_area - AssertionError: 'r1' wrong number of route nexthops
> FAILED isis_srv6_te_topo1/test_zx.py::test_srv6_te_policy_additional_route - AssertionError: "rt1" JSON output mismatches the expected result
> ================================================= 29 failed, 1732 passed, 279 skipped in 819.59s (0:13:39) =============================```
> 
> Please spend some time running your new test under different loads and identifying and fixing the problems found. 
> ```


Hi @donaldsharp , thanks for this comment. Are you running the topotests on the upstream CI hardware?
Reproduction attempt on my local machine of the concurrent test (including parallel load and many tests) run,  doesn't face this failures.

```
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.2.2, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/frr/frr/tests/topotests
configfile: pytest.ini
plugins: xdist-3.6.1
created: 12/12 workers
12 workers [7206 items]

scheduling tests via LoadFileScheduling

=============================== warnings summary ===============================
..
[gw8] PASSED isis_srv6_te_topo1/test_isis_srv6_te_topo1af.py::test_srv6_te_policy_removed 
isis_srv6_te_topo1/test_isis_srv6_te_topo1af.py::test_memory_leak 
[gw1] SKIPPED isis_srv6_te_topo1/test_isis_srv6_te_topo1ag.py::test_memory_leak 
[gw5] PASSED isis_srv6_te_topo1/test_isis_srv6_te_topo1ai.py::test_srv6_te_policy_removed 
isis_srv6_te_topo1/test_isis_srv6_te_topo1ai.py::test_memory_leak 
[gw7] SKIPPED isis_srv6_te_topo1/test_isis_srv6_te_topo1ah.py::test_memory_leak 
[gw8] SKIPPED isis_srv6_te_topo1/test_isis_srv6_te_topo1af.py::test_memory_leak 
[gw5] SKIPPED isis_srv6_te_topo1/test_isis_srv6_te_topo1ai.py::test_memory_leak 
[gw9] PASSED bgp_local_asn_dot/test_bgp_local_asn_dot_topo1.py::test_verify_bgp_local_as_allow_as_in_iBGP_p0 
bgp_local_asn_dot/test_bgp_local_asn_dot_topo1.py::test_verify_bgp_local_as_in_EBGP_port_reset_p0 
[gw6] PASSED multicast_pim_sm_topo1/test_multicast_pim_sm_topo1.py::test_modify_igmp_query_timer_p0 
multicast_pim_sm_topo1/test_multicast_pim_sm_topo1.py::test_modify_igmp_max_query_response_timer_p0 
isis_srv6_te_topo1/test_isis_srv6_te_topo1al.py::test_isis_adjacencies_step1 
isis_srv6_te_topo1/test_isis_srv6_te_topo1am.py::test_isis_adjacencies_step1 
isis_srv6_te_topo1/test_isis_srv6_te_topo1an.py::test_isis_adjacencies_step1 
isis_srv6_te_topo1/test_isis_srv6_te_topo1ao.py::test_isis_adjacencies_step1 
[gw0] PASSED multicast_pim_uplink_topo1/test_multicast_pim_uplink_topo1.py::test_mroutes_updated_correctly_after_source_interface_shut_noshut_p1 
multicast_pim_uplink_topo1/test_multicast_pim_uplink_topo1.py::test_mroutes_updated_correctly_after_receiver_interface_shut_noshut_p1 
[gw3] PASSED ospfv3_basic_functionality/test_ospfv3_authentication.py::test_ospf6_auth_trailer_tc8_keychain_sha256_missmatch 
ospfv3_basic_functionality/test_ospfv3_authentication.py::test_ospf6_auth_trailer_tc9_keychain_not_configured 
[gw4] PASSED isis_srv6_te_topo1/test_isis_srv6_te_topo1aj.py::test_isis_adjacencies_step1 
isis_srv6_te_topo1/test_isis_srv6_te_topo1aj.py::test_configure_srv6_locators 
[gw4] PASSED isis_srv6_te_topo1/test_isis_srv6_te_topo1aj.py::test_configure_srv6_locators 
isis_srv6_te_topo1/test_isis_srv6_te_topo1aj.py::test_rib_ipv6_step1 
[test.txt](https://github.com/user-attachments/files/16660557/test.txt)

..
========= 6399 passed, 808 skipped, 10 warnings in 4328.23s (1:12:08) ==========
```

Best Dmytro

dmytroshytyi-6WIND commented:
> as a note when I run the test by itself or concurrently with itself it works fine. It's only when scale/load is placed on the system that causes it to fail badly. This means, to me, that the test is starting to look for data before some other event has occurred and we get this output. See the hundreds of fixes in the topotest directory as we fix this sort of problem in the test for guidance on how to approach the problem. #16586, #16583, #16523, #16510 and #16485 as recent examples of this sort of problem.

Hello @donaldsharp , I have checked all PRs you suggested:

#16586 :
```
uses topotest.run_and_expect(..) to ensure that Connected Routes are actually installed
timeout 30
```

#16586 :
This test uses the same tool but with bigger timeout, so let's stick to higher timeouts then.
```
topotest.run_and_expect(..) +output = net["r1"]
timeout 60
```

#16523 :
This test introduce different way to use show. Two previous (majority stick to run_and_expect (so run_and_expect is selected))
```
@retry(retry_timeout=10)
def check_client_connect(r1):
    out = r1.vtysh_cmd("show mgmt backend-adapter all")
    assert "mgmtd-testc" in out
```

#16510:
To the TE policy route installed on local machine timers are not related
```
Timeout 60.
Sets timers to BGP 
```

#16485:

```
Addresses flapping routes, not sure if the current checks are a subject of being changed in this direction. They conform with firs two topotests.
```


Summary:
-+-+-+-+-

Timeout was set to 120 for all tests in this PR.

The difference I see between this PR and listed in quoted message - the way to configure the vtysh:
this PR is not using:
```
 output = net["r1"].cmd(
            'vtysh -c "show ip route {} nexthop-group"'.format(route_str)
        )
```
but instead uses:
```
    tgen.gears["rt1"].vtysh_cmd(
        "configure \n \
        no ipv6 route 2001:db8:10::/64 fc00:0:6:: color 1 \n \
        ipv6 route fc00:0:6b::/48 fc00:0:6:: color 1 \n \
        exit \
        !"
    )
```

Has someone encountered similar behavior when configuring in this way? I would do and verify the change, but  have a setup that doesn't reproduce mentioned issue (regarding the failed tests under the load).

dmytroshytyi-6WIND commented:
Failures are unrelated:
1. IPv4:
```
RFC Failure: MUST

Peer 192.168.0.101 did not forward MPLS packet with label 17
============== Reference: ============================
RFC 3036, s2.7 p23 LDP Identifiers and Next Hop Addresses
============== Test Summary: =========================
Similarly, when the LSR learns a label for a prefix from an LDP peer,
it must be able to determine whether that peer is currently a next hop
(3 more lines...)
```
2. ospf
```
test_ospfv3_single_area test_ospfv3_hello_tc10_p0 | 1 min
AssertionError: Testcase test_ospfv3_hello_tc10_p0 [22]: Failed     Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1 assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 [22]: Failed         Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1     assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
1 min
AssertionError: Testcase test_ospfv3_hello_tc10_p0 [22]: Failed 
   Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
E   AssertionError: Testcase test_ospfv3_hello_tc10_p0 [22]: Failed 
       Error: [DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1
    assert '[DUT: r0] missing OSPF neighbor r1 with router-id 100.1.1.1' is True
```

riw777 commented:
I think we're still waiting on @donaldsharp to clear his review on this one ... and the CI changes ... has this been rebased in a while? It might help to rebase it

dmytroshytyi committed:
yang,pathd: add srv6-sid-value and nb handlers

Add SRv6 SID value (srv6-sid-value) to the YANG model.
Implement nb handlers for this new value.

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

riw777 commented:
CI failures look related ...

dmytroshytyi-6WIND commented:
Yes, it is related , after rabase.
I will take a look.

Dmytro

dmytroshytyi-6WIND commented:
ci:rerun

dmytroshytyi-6WIND commented:
ci:rerun

dmytroshytyi-6WIND commented:
I fixed this PR:

@donaldsharp, @riw777 , @ton31337,

Could you please take a look one more time?

riw777 reviewed (approved):
this looks good to me at this point ... need to get the other blockers off ...

dmytroshytyi-6WIND commented:
ci:rerun

riw777 commented:
@frrbot autoclose in 2 weeks
